app/scripts/controllers/permissions/methodMiddleware.js

```diff
@@ -21,6 +29,19 @@ export default function createMethodMiddleware ({
 
       case 'eth_requestAccounts':
 
+        if (isProcessingRequestAccounts) {
+          res.error = ethErrors.rpc.resourceUnavailable(
+            'Already processing eth_requestAccounts. Please wait.'
```

Comment:
Is this case needed? I thought we were already preventing this in #8148

Reply:
Yeah, it's needed for when the extension is locked _and_ the `eth_accounts` permission is already granted.

If the permission is _not_ granted, the permission will be requested a pending approval will be created. If an approval is already pending, any additional requests will be rejected. This is what #8148 does.

Without this PR, if the permission _is_ granted _and_ the extension is locked, this method breaks. Why? Because `let accounts = await getAccounts()` will return `[]`, and then create a pending approval. This will re-request the permission completely, even though it's already granted, and should just return the already-approved accounts.

In short, without this PR, this method can't tell whether `getAccounts()` returns `[]` because the extension is locked or because the permission isn't granted. This PR allows this method to distinguish between these cases and behave appropriately in both.

Reply:
Right - I just meant to question whether this `if (isProcessingRequestAccounts) {` branch was needed. I had thought the `await getUnlockedPromise()` below would solve the problem you're describing.

Though I guess #8148 wouldn't stop these unlock requests from piling up either, so maybe this second-layer of preventing duplicate requests is necessary 🤔

Reply:
Ah, I see.

And yeah, that statement is exactly to prevent unlock requests from piling up.